Parametric Risch differential equation

jksuom opened the pull request:
This contains an implementation of the high level functions for the solution of parametric Risch differential equations.

Some low level routines for cancellation cases are not yet implemented.

jksuom committed:
Create poly_linear_constraints

Simplified version of linear_constraints for polynomials.

jksuom committed:
Create param_poly_rischDE

jksuom committed:
Create param_rischDE

Add tests.

jksuom committed:
Extension of prde_no_cancel_b_small

asmeurer commented:
I haven't been able to review all the algorithms in depth, but what I have looked at looks good. The real test will be integrating this into `limited_integrate` [here](https://github.com/sympy/sympy/blob/4279fb3fb7db5cfe27d999e6e3154575d9a0aa3c/sympy/integrals/prde.py#L455).

gxyd commented on sympy/integrals/prde.py:
```diff
@@ -188,11 +188,29 @@ def prde_linear_constraints(a, b, G, DE):
     qs, _ = list(zip(*Q))
     return (qs, M)
 
+def poly_linear_constraints(p, d):
```
Is this function an improvement over the [prde_linear_constraints](https://github.com/sympy/sympy/blob/master/sympy/integrals/prde.py#L166)?

gxyd committed:
Add call to param_rischDE in limited_integrate

Call param_rischDE on top of limited_integrate

test with `not V` and remove the assertion that len(V) == 1

gxyd commented on sympy/integrals/prde.py:
```diff
@@ -755,56 +755,65 @@ def limited_integrate_reduce(fa, fd, G, DE):
+    V = A.nullspace()
+    V = [v for v in V if v[0] != 0]
+    if V is None:
+        return None
```
I think this should raise `NonElementaryIntegral` exception instead of returning `None`?

jksuom replied:
We may eventually want to handle also nonelementary functions. `limited_integrate` may then be called from different places in the code, and it should be up to the caller to decide whether to raise an exception or do something else. We should consider `limited_integrate` as a low level routine that only answers the question whether there is an integral in the limited sense or not.

gxyd replied:
So it is upto the caller to decide that in case of a non-elementary integral whether to straight away raise an error (`NonElementaryIntegralException`) or to return the answer in terms of common non-elementary functions (like `Erf`, `Ei`, `Li` etc.).

gxyd replied:
I do remember the conversation we had regarding what "limited" means, so everything is clear.

asmeurer commented on sympy/integrals/prde.py:
```diff
@@ -437,40 +744,76 @@ def limited_integrate_reduce(fa, fd, G, DE):
+    h, A = param_rischDE(Fa, Fd, G, DE)
+    V = A.nullspace()
+    V = [v for v in V if v[0] != 0]
+    if V is None:
```
Should be `if not V` (`V` is a list and won't be None).

asmeurer commented on sympy/integrals/prde.py:
```diff
@@ -437,40 +744,76 @@ def limited_integrate_reduce(fa, fd, G, DE):
+    G = [(fa, fd)] + G
+    h, A = param_rischDE(Fa, Fd, G, DE)
+    V = A.nullspace()
+    V = [v for v in V if v[0] != 0]
```
From the book it seems like `v[0] == 0` or even `len(V) > 1` should indicate a bug. Can you verify if that is the case from examples?

jksuom replied:
I think that those lines in the book only refer to some (possibly quite common) cases where various shortcuts can be taken.

asmeurer commented on sympy/integrals/prde.py:
```diff
@@ -437,40 +744,76 @@ def limited_integrate_reduce(fa, fd, G, DE):
     else:
-        raise NotImplementedError
+        c0 = V[0][0]
+        # only one vector expected in V
```
Should check this condition with an assert.

gxyd commented on sympy/integrals/prde.py:
```diff
@@ -442,35 +749,29 @@ def limited_integrate(fa, fd, G, DE):
+    V = A.nullspace()
+    V = [v for v in V if v[0] != 0]
+    if not V:
+        return NonElementaryIntegralException
```
Okay, made a mistake again, it should be `raise`, will change it.

jksuom replied:
Can we say this already at this point? Maybe the caller should decide.

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -442,35 +749,29 @@ def limited_integrate(fa, fd, G, DE):
     else:
-        raise NotImplementedError
+        c0 = V[0][0]
+        V = [v/(-c0) for v in V]
```
It is not necessary to divide all vectors in `V`. It is enough to take and normalize the first one:

    v = V[0]/(-c0)

gxyd committed:
Normalize only one vector and return None in case of no solution

raise NonElementaryIntegralException; fix Poly/Poly error

On call of limited_integrate by limited_integrate_polynomial raise
NonElementaryIntegralException when limited_integrate returns None

Add import of `oo` and remove check of len(c) != 1

Fix docstring and add a #TODO with removed indent

jksuom commented on sympy/integrals/risch.py:
```diff
@@ -1283,7 +1283,10 @@ def integrate_primitive_polynomial(p, DE):
+                if rv is None:
+                    raise NonElementaryIntegralException
+                (ba, bd), c = rv
                 if len(c) != 1:
```
There is probably no reason to have this test in production code as the argument vector has length 1.

gxyd replied:
Oh, of course. As for this case `g1, g2`, will be there => `c1, c2` with c1 = 1 (fixed) we get v = [c2] hence a fixed length.

gxyd committed:
use PolyMatrix instead of Matrix in prde.py and test file

gxyd commented:
In master:
```
gxyd@stallman:~/foss/sympy$ ipython3
Python 3.6.0 |Anaconda 4.3.1 (64-bit)| (default, Dec 23 2016, 12:22:00) 
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from sympy import *

In [2]: from sympy.integrals.risch import risch_integrate

In [3]: x = Symbol('x')

In [4]: risch_integrate(log(x + 1)/log(x), x)
Out[4]: Integral(log(x + 1)/log(x), x)

In [5]: from sympy.integrals.risch import risch_integrate, NonElementaryIntegral

In [6]: isinstance(risch_integrate(log(x + 1)/log(x), x), NonElementaryIntegral)
Out[6]: True    # this raises `prde_no_cancel` not implemented error in param-rde
```

jksuom commented:
On which line (or which case)?

gxyd commented:
```
In [4]:  risch_integrate(log(x)*log(x + 1), x)
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-4-fc5e78bbc6b2> in <module>()
----> 1 risch_integrate(log(x)*log(x + 1), x)

/home/gxyd/foss/sympy/sympy/integrals/risch.py in risch_integrate(f, x, extension, handle_first, separate_integral, rewrite_complex, conds)
   1690             ans, i, b = integrate_hyperexponential(fa, fd, DE, conds=conds)
   1691         elif case == 'primitive':
-> 1692             ans, i, b = integrate_primitive(fa, fd, DE)
   1693         elif case == 'base':
   1694             # XXX: We can't call ratint() directly here because it doesn't

/home/gxyd/foss/sympy/sympy/integrals/risch.py in integrate_primitive(a, d, DE, z)
   1337     p = p.as_poly(DE.t)
   1338 
-> 1339     q, i, b = integrate_primitive_polynomial(p, DE)
   1340 
   1341     ret = ((g1[0].as_expr()/g1[1].as_expr() + q.as_expr()).subs(s) +

/home/gxyd/foss/sympy/sympy/integrals/risch.py in integrate_primitive_polynomial(p, DE)
   1289 
   1290             try:
-> 1291                 rv = limited_integrate(aa, ad, [(Dta, Dtb)], DE)
   1292                 if rv is None:
   1293                     raise NonElementaryIntegralException

/home/gxyd/foss/sympy/sympy/integrals/prde.py in limited_integrate(fa, fd, G, DE)
    757     Fd = Poly(1, DE.t)
    758     G = [(fa, fd)] + G
--> 759     h, A = param_rischDE(Fa, Fd, G, DE)
    760     V = A.nullspace()
    761     V = [v for v in V if v[0] != 0]

/home/gxyd/foss/sympy/sympy/integrals/prde.py in param_rischDE(fa, fd, G, DE)
    657         n = oo
    658 
--> 659     h, B = param_poly_rischDE(a, b, r, n, DE)
    660 
    661     # h = [h1, ..., hv] in k[t]^v and and B is a matrix with u + v

/home/gxyd/foss/sympy/sympy/integrals/prde.py in param_poly_rischDE(a, b, q, n, DE)
    487         else:
    488             # Cancellation
--> 489             raise NotImplementedError("prde_cancel() is not yet "
    490                 "implemented.")
    491 

NotImplementedError: prde_cancel() is not yet implemented.
```

jksuom commented:
So it is prde_cancel, not prde_no_cancel. That looks like the place where the 'Liouvillian case' on p. 237 should be called. That should be the first missing piece to implement. The nonlinear and hypertangent cases are not so important yet.

gxyd commented:
Yes it should be `prde_cancel` not `prde_no_cancel`, I have edited the comment.

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -441,6 +441,52 @@ def prde_no_cancel_b_small(b, Q, n, DE):
+        # fn = f1n, ..., frn_n
+        # An is a matrix with coefficients in Const(k)
+        # yn = Sum(djn*fjn, (j, 1, rn))
+        fn, An = param_rischDE(ba, bd, Qyn, DE)
```
`param_rischDE` should be called repeatedly for descending degrees n, n -1, ..., 0, and the results should probably be saved in a list like `f[n]`, `A[n]` to be combined at the end before returning. Alternatively, it might be possible to save some list space by doing a partial combination of the results after each call of `param_rischDE`.

gxyd commented on sympy/integrals/prde.py:
```diff
@@ -359,35 +379,381 @@ def prde_no_cancel_b_small(b, Q, n, DE):
+
+    if a.is_ground:
+        # Normalization: a = 1.
+        a = a.LC()
```
I am not completely aware of complete algorithm, but this seems like a silent change from a `Poly` instance to a `Number`. Then when it is called on [this line](https://github.com/sympy/sympy/pull/11761/files#diff-25c3a80e410bd47d98bfac9be56b42b1R555) it raises `AttributeError: 'One' object has no attribute 'degree'`. Otherwise change from `Matrix` to `PolyMatrix` fits nicely.

jksuom replied:
Which line is that?

gxyd replied:
This one https://github.com/sympy/sympy/pull/11761/files#diff-25c3a80e410bd47d98bfac9be56b42b1R555 i.e `prde_spde`.
```python
[gxyd@stallman sympy{param-rde}]$ python3 ./bin/test sympy/integrals/tests/test_prde.py
============================================================= test process starts =============================================================
executable:         /home/gxyd/anaconda3/bin/python3  (3.6.0-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        66232951
hash randomization: on (PYTHONHASHSEED=2396363264)

sympy/integrals/tests/test_prde.py[15] ..............E                                                                                   [FAIL]

_______________________________________________________________________________________________________________________________________________
_____________________________________________ sympy/integrals/tests/test_prde.py:test_issue_10798 _____________________________________________
  File "/home/gxyd/foss/sympy/sympy/integrals/tests/test_prde.py", line 278, in test_issue_10798
    assert integrate(1/(1-(x*y)**2), (x, 0, 1), y) == \
  File "/home/gxyd/foss/sympy/sympy/integrals/integrals.py", line 1295, in integrate
    risch=risch, manual=manual)
  File "/home/gxyd/foss/sympy/sympy/integrals/integrals.py", line 486, in doit
    conds=conds)
  File "/home/gxyd/foss/sympy/sympy/integrals/integrals.py", line 780, in _eval_integral
    result, i = risch_integrate(f, x, separate_integral=True, conds=conds)
  File "/home/gxyd/foss/sympy/sympy/integrals/risch.py", line 1692, in risch_integrate
    ans, i, b = integrate_primitive(fa, fd, DE)
  File "/home/gxyd/foss/sympy/sympy/integrals/risch.py", line 1339, in integrate_primitive
    q, i, b = integrate_primitive_polynomial(p, DE)
  File "/home/gxyd/foss/sympy/sympy/integrals/risch.py", line 1291, in integrate_primitive_polynomial
    rv = limited_integrate(aa, ad, [(Dta, Dtb)], DE)
  File "/home/gxyd/foss/sympy/sympy/integrals/prde.py", line 816, in limited_integrate
    h, A = param_rischDE(Fa, Fd, G, DE)
  File "/home/gxyd/foss/sympy/sympy/integrals/prde.py", line 716, in param_rischDE
    h, B = param_poly_rischDE(a, b, r, n, DE)
  File "/home/gxyd/foss/sympy/sympy/integrals/prde.py", line 555, in param_poly_rischDE
    a, b, q, r, n = prde_spde(a, b, q, n, DE)
  File "/home/gxyd/foss/sympy/sympy/integrals/prde.py", line 306, in prde_spde
    n1 = n - a.degree(DE.t)
AttributeError: 'One' object has no attribute 'degree'
```

jksuom replied:
It seems that something goes wrong in the construction of DE before any integration is attempted:
```
f = (log(-1/y)/2 - log(1/y)/2)/y - (log(1 - 1/y)/2 - log(1 + 1/y)/2)/y

In [4]: pdb.run('risch_integrate(f, y)')
> <string>(1)<module>()
(Pdb) s
--Call--
> /home/jks/test/sympy/sympy/integrals/risch.py(1570)risch_integrate()
-> def risch_integrate(f, x, extension=None, handle_first='log',
(Pdb) 
> /home/jks/test/sympy/sympy/integrals/risch.py(1675)risch_integrate()
-> f = S(f)
(Pdb) n
> /home/jks/test/sympy/sympy/integrals/risch.py(1677)risch_integrate()
-> DE = extension or DifferentialExtension(f, x, handle_first=handle_first,
(Pdb) 
> /home/jks/test/sympy/sympy/integrals/risch.py(1678)risch_integrate()
-> dummy=True, rewrite_complex=rewrite_complex)
(Pdb) 
> /home/jks/test/sympy/sympy/integrals/risch.py(1679)risch_integrate()
-> fa, fd = DE.fa, DE.fd
(Pdb) p DE
DifferentialExtension(extension=dict(fa=Poly(-2*y*_t2 + 2*_t1*y + 2*I*pi*y, _t2, domain='EX'), fd=Poly(4*y**2, _t2, domain='ZZ[y]'), D=[Poly(1, y, domain='ZZ'), Poly(-1/y, _t0, domain='ZZ(y)'), Poly(-1/(y**2 + y), _t1, domain='ZZ(y)'), Poly(1/(y**2 - y), _t2, domain='ZZ(y)')]))
```
The domain of `fa` should not be `EX` but rather something like `ZZ(pi, y)`. I'll try to see how that will affect the workings of prde.

gxyd replied:
Yes, the generator is explicitly specified to be `_t2` not `_t2, I`.

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -359,35 +379,381 @@ def prde_no_cancel_b_small(b, Q, n, DE):
+                "not yet implemented.")
+        else:
+            if b.is_zero:
+                pass
```
This should be something else. It now leads to line 549, where it is assumed that `a` is not constant.

jksuom replied:
`a.LC()` may not be optimal here since it sympifies the coefficient. `a.nth(0)` does the same. 
But it does not matter since `quo_ground` will convert `a` back anyway. (Only some time is lost in the conversions.) And `a` should not be used elsewhere.

gxyd commented on sympy/integrals/prde.py:
```diff
@@ -359,35 +379,388 @@ def prde_no_cancel_b_small(b, Q, n, DE):
+    # over k. That is why we are using DecrementLevel.
+    with DecrementLevel(DE):
+        t0 = DE.t
+        ba, bd = frac_in(b, t0)
```
I put this out of the loop, since it is just going to remain same. Only review this part and the first loop below it for now.

gxyd commented on sympy/integrals/tests/test_prde.py:
```diff
@@ -1,11 +1,13 @@
-    is_log_deriv_k_t_radical_in_field)
+    is_log_deriv_k_t_radical_in_field, param_poly_rischDE, param_rischDE)
+
+from sympy.polys.polymatrix import PolyMatrix
```
A complete change from `Matrix` to `PolyMatrix` will do in the next PR.
